### ISSUE TITLE
Faster checkout process using GeoIP and IPN billing data

### ISSUE THREAD

**Issue by davidsherlock** [easydigitaldownloads/easy-digital-downloads#5666]
I recently visited a digital download store that used a combination of GeoIP and PayPal IPN billing data to make the entire checkout process quicker and smoother for the customer. Effectively this done away with needing "billing details" on the site itself and simply presented fields where you entered your email address, name and any discount codes then went to PayPal and paid your order. PayPal then sent back your account billing details as per of the IPN callback, which resulted in tax getting properly calculated and invoices generated using the PayPal billing details (while it appears creating a customer record in the background).

For obvious reasons this results in the customer not having to enter their details twice and a much faster checkout process.

**Comment by pippinsplugins** [comment]
I've wanted to use GeoIP for tax calculations for a long time. It'd be a great improvement.

Ideally, we'd use GeoIP to detect the location and show the detected tax rates but then allow a customer to overwrite that with a manual address entry.

**Comment by pippinsplugins** [comment]
IPN data could potentially work but there are a lot of issues with calculating tax and recording it _after_ payment is completed. It needs to come before payment is submitted, so using IPN data for tax calculations won't really be doable.

**Comment by cklosowski** [comment]
The real big 'problem' (if you want to call it that) with GeoIP is the lack of good services that are affordable AND accurate.

Of the ones we use in Fraud Monitor, one fails frequently b/c it's crowdsourced, and the other only allows 2500 queries a day before charging. The other concern is speed/performance. We can find ways around this but something we need to keep in mind. API calls need to be able to fail gracefully.

We could see a great improvement here for sure, but we'll have to be mindful of the reliance on 3rd party services, unless we include a local DB file (which are large in size, in MBs).

**Comment by davidsherlock** [comment]
I agree Pippin, though my suggestion was more in line with optimizing (and speeding up) the checkout process and minimizing form filling/duplication. With the exception of tax, could IPN be used to fill in billing/customer details when returned to the site if the customer doesn't have an account? One issue I can forsee is if a customer is out of the country and GeoIP is used, the incorrect tax will be recorded which causes issues with EU VAT regulations in line with the place of supply rules.

One solution could be to simply have a 'slimmer' checkout form with only the country and post / zip code fields, then use GeoIP to initially "guess" these but let the customer override them via these fields.

**Comment by pippinsplugins** [comment]
>>>could IPN be used to fill in billing/customer details when returned to the site if the customer doesn't have an account?

We actually already do that, though there's probably some additional data we could update if not present.

I have no qualms with this requiring a paid service. It'd be an optional feature anyhow that is disabled by default.

**Comment by davidsherlock** [comment]
@cklosowski, slightly off-topic but does fraud monitor compare the IPN billing address (or even country) to the one entered as their "billing address" within EDD? Having a different country can be a big red flag for fraudulent orders.

**Comment by davidsherlock** [comment]
> I have no qualms with this requiring a paid service. It'd be an optional feature anyhow that is disabled by default.

I second this, for anyone serious about creating a great customer experience this would be something worth paying for. MaxMind is very affordable for this type of service.

**Comment by cklosowski** [comment]
@davidsherlock It doesn't use IPN data, it uses IP and billing data.

As of now it can flag for:
1) IP Country being flagged
2) IP Country not matching Billing country
3) IP Country and billing country being X miles/km away from each other
4) File download coming from banned country (after purchase)
5) File download coming from a different country than purchase within 24 hours of the purchase completion

Amongst other things, these are just the GeoIP location ones.

**Comment by davidsherlock** [comment]
@cklosowski, for the sake of clarity - you mean the EDD billing data and not the data PayPal sends back?

I am going to open a thread on the fraud monitor repository, but it might be worth also checking the headers for proxy/VPN usage as well (out with just IP checks).

**Comment by davidsherlock** [comment]
user-agent checking would also be useful.

**Comment by cklosowski** [comment]
@davidsherlock You can open one up there, but the current solutions were put in place after trying to determine the proxy/VPN usage and running into many roadblocks  and false positives.

**Comment by davidsherlock** [comment]
Chris, an additional check could be:

PayPal billing address country and EDD billing country being X miles/km away from each other, or a mis-match.

**Comment by davidsherlock** [comment]
@pippinsplugins, couldn't the GeoIP tax calculation functionality allow Stripe "buy now" to work as intended if implemented?

**Comment by pippinsplugins** [comment]
@davidsherlock 
>>>couldn't the GeoIP tax calculation functionality allow Stripe "buy now" to work as intended if implemented?

Possibly. It would sure make a lot of people happy if so!
